### PR TITLE
8303669: SelectVersion indexes past the end of the argv array

### PR DESCRIPTION
**Pull request by asotona**
libjli/java.c's SelectVersion method receives argc and argv but ignores argc in some circumstances an instead checks if *argv == 0 in its while loop, which results in a segmentation fault if the provided array is not NULL terminated. 

This patch counts down argc in the while loops instead of looking for zero termination.

Please review.

Thank you,
Adam

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303669](https://bugs.openjdk.org/browse/JDK-8303669): SelectVersion indexes past the end of the argv array

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13775/head:pull/13775` \
`$ git checkout pull/13775`

Update a local copy of the PR: \
`$ git checkout pull/13775` \
`$ git pull https://git.openjdk.org/jdk.git pull/13775/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13775`

View PR using the GUI difftool: \
`$ git pr show -t 13775`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13775.diff">https://git.openjdk.org/jdk/pull/13775.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13775#issuecomment-1532915581)

### CODE REVIEW

File: src/java.base/share/native/libjli/java.c [hunk 1]
```diff
@@ -1209,7 +1209,7 @@ ParseArguments(int *pargc, char ***pargv,
 
     *pret = 0;
 
-    while ((arg = *argv) != 0 && *arg == '-') {
+    while (argc > 0 && *(arg = *argv) == '-') {
```
Comment (dholmes-ora) [inline src/java.base/share/native/libjli/java.c:1212]: AFAICS this loop terminates at line 1388 and nowhere in the loop body does argc get modified. ??

**Comment by AlanBateman** [comment]
I think this bug report is a reminder that we need to figure out if JLI was documented at some point, it's surprising to have something outside of the JDK using it directly.

Also are you planning to add a test for this?

File: src/java.base/share/native/libjli/java.c (hunk 1 above)
Comment (asotona) [reply src/java.base/share/native/libjli/java.c:1212]: Reference to argc is passed down to GetOpt, where it is updated.

**Comment by asotona** [comment]
> Also are you planning to add a test for this?

I've added code to `exeJliLaunchTest.c` (executed by `JliLaunchTest`) to test non null terminated arguments.

File: src/java.base/share/native/libjli/java.c (hunk 1 above)
Comment (dholmes-ora) [reply src/java.base/share/native/libjli/java.c:1212]: Ah I see - ouch!

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

**Comment by asotona** [comment]
Thank you.

**Pull request closed without merging**